src/front/glsl/builtins.rs

```diff
@@ -118,31 +138,83 @@ pub fn inject_builtin(declaration: &mut FunctionDeclaration, module: &mut Module
+                    // textureProjGrad(gsampler, gvec+1 P, gvec dPdx, gvec dPdy);
+                    "textureProjGrad" => MacroCall::Texture { proj: true, offset: false, level_type: TextureLevelType::Grad(shadow), },
+                    // textureProjGradOffset(gsampler, gvec+1 P, gvec dPdx, gvec dPdy, ivec offset);
+                    "textureProjGradOffset" => MacroCall::Texture { proj: true, offset: true, level_type: TextureLevelType::Grad(shadow), },
+                    // textureProjLod(gsampler, gvec+1 P, float lod);
+                    "textureProjLod" => MacroCall::Texture { proj: true, offset: false, level_type: TextureLevelType::Lod, },
+                    // textureProjLodOffset(gsampler, gvec+1 P, gvec dPdx, gvec dPdy, ivec offset);
+                    "textureProjLodOffset" => MacroCall::Texture { proj: true, offset: true, level_type: TextureLevelType::Lod, },
+                    // textureProjOffset(gsampler, gvec+1 P, ivec offset, [float bias]);
+                    "textureProjOffset" => MacroCall::Texture { proj: true, offset: true, level_type: TextureLevelType::None, },
+                    _ => unreachable!(),
                 };
 
-                // Shadow, arrayed or both 3D images are not allowed
-                if (shadow || arrayed) && dim == 0b11
-                    || (builtin == MacroCall::TextureLod
-                        && (dim != 0b00 && shadow && arrayed || dim == 0b10 && shadow || !bias))
-                {
+                // Parse out the variant settings.
+                let proj = matches!(builtin, MacroCall::Texture { proj: true, .. });
+                let grad = matches!(builtin, MacroCall::Texture { level_type: TextureLevelType::Grad(_), .. });
+                let lod = matches!(builtin, MacroCall::Texture { level_type: TextureLevelType::Lod, .. });
+                let offset = matches!(builtin, MacroCall::Texture { offset: true, .. });
+
+                if variant {
+                    let supports_variant = proj && !shadow;
+                    if !supports_variant {
+                        continue;
+                    }
+                }
+
+                if bias {
+                    let supports_bias = matches!(builtin, MacroCall::Texture { level_type: TextureLevelType::None, .. }) && !shadow;
+                    if !supports_bias {
+                        continue;
+                    }
+                }
+
+                // Proj doesn't work with arrayed, Cube or 3D samplers
+                if proj && (arrayed || dim == 0b10 || dim == 0b11) {
+                    continue;
+                }
+
+                // 3DArray and 3DShadow are not valid texture types
+                if dim == 0b11 && (arrayed || shadow) {
+                    continue;
+                }
+
+                // Naga IR restriction -- explicit LOD is not allowed on shadow samplers.
+                if lod && shadow {
```

Comment:
Check how spv-in handles this https://github.com/gfx-rs/naga/blob/8aae069f22ac94ec46329a9adbfaf341cd3e7adf/src/front/spv/image.rs#L455-L467

It assumes `Zero` and throws an warning, we should do the same

Reply:
I think there's code to do that, but it's not quite working, let me look into why.

Reply:
Ah, it was only set up to work on Grad, it never worked on Lod. Fixed.